### ISSUE TITLE
matplotlib.widget.TextBox not correctly stopping keyboard shortcuts

### ISSUE THREAD

**Issue by mikehenninger** [matplotlib/matplotlib#7571]
I found that even with the attempt to turn off shortcuts when entering text in textbox, they still had key presses interpreted as the shortcuts, at least with the tkagg backend. I have what appears to be a working solution (store the callback, call self.canvas.callbacks.disconnect(cid) to completely disconnect everything looking for keystrokes in that axes in begin_typing(), then reconnect the callbacks in stop_typing(), but I don't have time to submit pull request right now. This would be my first time submitting to any open source project, much less MPL. Hell, I haven't even used GIT on a project involving more than just myself. I thought it'd be good to throw it out as an issue before I got ahead of myself.

-using tkagg backend
-using latest (yesterday's) master branch to test out TextBox widget.
-using winpython 64bit 3.5.2.3QT5, except for MPL which I built from source

**Comment by anntzer** [comment]
As of master, I actually observe a different problem, both with tkagg and qt5agg (on Linux, which may be the difference): the shortcuts are correctly inactive while typing, but they stay inactive even after leaving the textbox (`examples/widgets/textbox.py`).

**Comment by mikehenninger** [comment]
Ah yes I think I might know the cause of that as well. I'm not at my
computer but there was an instance of erroneously turning off an is typing
flag before calling stop_typing. I'll see if I can indeed repro tomorrow
and confirm that's the fix, then hopefully do a pull request with fixes for
both our issues.

 I'm mostly worried about side effects of my fix for the original bug,
given I don't know the code base very well...

**Comment by anntzer** [comment]
Give it a try and we'll look at it :)

**Comment by mikehenninger** [comment]
@anntzer, I reproduced your bug and fixed it;  was indeed a bad flag flip. One line change. I'll submit a pull request for that. 

As far as my original issue, it's proven to be toolmanager-related: 
Iff you set rcParams['toolbar'] = 'toolmanager', Qt5Agg keyboard shortcuts never work, regardless of presence of TextBox. And, as I originally noted, with TkAgg they don't disable properly in the textbox. I haven't tested other backends. While I do have a fix for the TkAgg version of the problem, it seems like a band-aid solution that doesn't address whatever the real problem is. I don't want to wade into the backend code just yet, and applying a backend-specific fix in widgets.py seems pretty bad.


```
"""demo of shortcut keys not behaving in textbox when used with toolmanager

TKAgg has never-disabled shortcut keys, Qt5Agg has a separate bug of no
shortcut keys work ever anywhere independently of textbox, other backends not
tested.

"""
import numpy as np
import matplotlib as mpl

# bug manifests when using both TkAgg and toolmanager
mpl.use('TkAgg')
mpl.rcParams['toolbar'] = 'toolmanager'
import matplotlib.pyplot as plt
from matplotlib.widgets import TextBox
fig, ax = plt.subplots()
plt.subplots_adjust(bottom=0.2)
t = np.arange(-2.0, 2.0, 0.001)
s = t ** 2
initial_text = "t ** 2"
l, = plt.plot(t, s, lw=2)


def submit(text):
    # note that this barfs (silently on some backends) on any not-math
    ydata = eval(text)
    l.set_ydata(ydata)
    ax.set_ylim(np.min(ydata), np.max(ydata))
    plt.draw()

axbox = plt.axes([0.1, 0.05, 0.8, 0.075])
text_box = TextBox(axbox, 'Evaluate', initial=initial_text)
text_box.on_submit(submit)

plt.show()

```

**Comment by anntzer** [comment]
attn @fariza re: toolmanager.

I think we should make an effort either to switch all backends to using toolmanager (probably a good idea) or just get rid of that code (at least for now), there isn't much point of having it lying around mostly unused.

**Comment by fariza** [comment]
Porting the code to all backends is a big amount of work, specially if it
is not stable/tested enough
We have decided that the code there (Gtk3 and Tk) is enough for people to
test.

The final conversion of the other backends will be done at the same time of
MEP27. It will be high priority after the current release is done.

I know it seems like dead code but the work keeps going on. We have done
several corrections and additions that are slowly integrated. I really hope
we will accelerate once this release is done.

Be patient, testers and tool developers are welcome

Federico

**Comment by tacaswell** [comment]
toolmanager stays

**Comment by anntzer** [comment]
(I completely agree with the idea of toolmanager (the current system is not very extensible); my point was simply that we should not try to maintain two different tool systems in parallel.)

**Comment by andres-fr** [comment]
Just in case someone stumbles into this, or for better suggestions to my approach, I tell my workaround to this. I was having this issue on 3.0.2:

1. I activated `matplotlib.rcParams["toolbar"] = "toolmanager"` because I wanted to add some buttons to the toolbar.

2. Every time I tiped e.g. `s` anywhere in the window (including textbox) the save menu would pop up.

3. Updating (i.e. disabling) the values under `matplotlib.rcParams["keymap..."]` didn't seem to have any effect.

I found out that the tool manager has a "parallel" key listener, accessible through the following protected attribute: `fig.canvas.manager.toolbar.toolmanager._keys`. Setting this to an empty (or any custom) dictionary did the job for me.

I would be glad to hear if there is any more stable way to perform this workaround, to have custom toolbars AND custom textboxes.  In any case hope this helps someone!  

Andres

**Comment by fariza** [comment]
@andres-fr did you try with the `toolmanager.keypresslock` if you acquire that lock, the tools should not be triggered

**Comment by fariza** [comment]
@andres-fr please test and let us know so we can close this issue

**Comment by andres-fr** [comment]
what do you mean with acquiring? I don't understand what should I try exactly

**Comment by fariza** [comment]
@andres-fr after close inspection to the textbox widget I made a small PR to fix the situation
https://github.com/matplotlib/matplotlib/pull/14343
If possible, please checkout that PR and confirm that fixes your bug

**Comment by andres-fr** [comment]
I am sorry but I lack experience testing PRs and am afraid to mess my setup, do you have any references on how to do that? I assume I would need a fresh venv

**Comment by fariza** [comment]
It's ok, no problem, testing PRs implies having your own matplotlib dev environment.
I tested with the basic textbox example and it works so it will work for you

**Comment by fariza** [comment]
As a temporary workaround that you can always do

```
class NewTextBox(TextBox):
   def begin_typing(self, x):
       super().begin_typing(x)
       self.ax.figure.canvas.manager.toolmanager.keypresslock(self)

    def stop_typing(self):
        super().stop_typing()
        self.ax.figure.canvas.manager.toolmanager.keypresslock.release(self)
```
And use this new class as you where using the normal TextBox, and once the fix gets merged and released you can just go back to use the TextBox class

**Comment by anntzer** [comment]
Closed by #14343, to the best of my understanding.
